src/Plugins/Reminder.php

```diff
@@ -0,0 +1,357 @@
+                $key  = base_convert($value['id'], 10, 36);
+                $timestamp = $value['timestamp'];
+
+                if (!$timestamp) {
+                    return $this->chatClient->postMessage($command->getRoom(), "Have a look at the time again, yo!");
+                }
+                $seconds = $timestamp - time();
+
+                if ($seconds <= 0) {
+                    $timeouts[] = $key;
+                    continue;
+                }
+
+                $message .= sprintf(
+                    "\n%s %s %s %s %s %s - %s - %s ",
+                    Chars::BULLET,
+                    $text,
+                    Chars::RIGHTWARDS_ARROW,
+                    "Id: " . $key,
+                    Chars::RIGHTWARDS_ARROW,
+                    date('l, dS F Y H:i (e)', $timestamp),
+                    'Set by ' . $user,
+                    'Seconds left: ' . $seconds
+                );
+            }
+
+            if(count($timeouts) !== count($reminders)){
+                return $this->chatClient->postMessage($command->getRoom(), $message);
+            }
+        });
+    }
+
+    /**
+     * Handle a command message
+     *
+     * According to http://www.strawpoll.me/11212318/r this plugin will support the following commands
+     * !!<reminder|remind> remind this <in|at> <time>
+     * !!at <time> remind this
+     * !!in <time> remind this
+     *
+     * @param Command $command
+     * @return Promise
+     */
+    public function handleCommand(Command $command): Promise
+    {
+        return resolve(function() use($command) {
+
+            if ($command->hasParameters() === false) {
+                return $this->chatClient->postMessage($command->getRoom(), self::USAGE);
+            }
+
+            /* $command->getParameter(0) can be: list | examples | flush | unset | <text> | <time> */
+            $textOrCommand = $command->getParameter(0);
+            $commandName = $command->getCommandName(); // <reminder|in|at>
+            switch ($commandName){
+                case 'reminder': case 'remind': break;
+                case 'in': case 'at': return yield $this->setReminder($command); break;
+            }
+
+            if ( count(array_diff($command->getParameters(), array($textOrCommand))) < 1 ){
+                switch ($textOrCommand){
+                    case 'list': return yield $this->getAllReminders($command); break;
+                    case 'examples': return yield $this->getExamples($command); break;
+                    case 'flush': return yield $this->flush($command);
+                }
+            }
+
+            if( $command->getParameter(0) === 'unset'
+                && $command->getParameter(1) !== null
+                && count($command->getParameters()) <= 2
+            ){ return yield $this->unset($command); }
+
+            return yield $this->setReminder($command);
+        });
+    }
+
+    public function apologize(ChatRoom $room, $timedOutReminders): \Generator
```

Comment:
Maybe rename to `apologizeExpired` and instead of pre-filtering `$timedOutReminders` just pass in all reminders and let this method find out what to handle.